### DISCUSSION TITLE
Mirth Sending TCP RST for Large Messages

### DISCUSSION THREAD

**Discussion by wcameronbowen** [nextgenhealthcare/connect#5617] [Q&A] [answered]
We currently have Mirth setup to receive realtime messages. We generally receive messages around 8KB in size. Periodically we receive a larger than normal message (~3MB). When this happens Mirth sends a TCP RST and closes the connection while receiving the message. 

<img width="1505" alt="mirth-packet-capture" src="https://user-images.githubusercontent.com/39662961/212974811-9a390965-b85f-4058-bf4f-30cf6dc78757.png">

The result is that we receive a lot of these errors in the server log. The last bytes correspond to the last PSH, ACK message from the upstream (source) message queue (in this case CloverLeaf).

```
[2023-01-11 01:09:26,852]  ERROR  (com.mirth.connect.connectors.tcp.TcpReceiver:746): Error receiving message (TCP Listener "Source" on channel f8bb636c-4414-4a9a-831e-5b8b0a340b26).
com.mirth.connect.model.transmission.framemode.FrameStreamHandlerException: End of message bytes (<FS><CR>) not detected. Last bytes received: 08
```

Enabling more DEBUG logs I see some timeouts reading from the socket, which may (or may not) be related.

```
DEBUG 2023-01-12 18:45:03,244 [TCP Receiver Thread on Notes (f8bb636c-4414-4a9a-831e-5b8b0a340b26) < pool-38-thread-1] com.mirth.connect.connectors.tcp.TcpReceiver: Bytes returned from socket, length: 8652 (TCP Listener "Source" on channel f8bb636c-4414-4a9a-831e-5b8b0a340b26)
DEBUG 2023-01-12 18:45:03,247 [TCP Receiver Thread on Notes (f8bb636c-4414-4a9a-831e-5b8b0a340b26) < pool-38-thread-1] com.mirth.connect.connectors.tcp.TcpReceiver: Done with socket input stream (TCP Listener "Source" on channel f8bb636c-4414-4a9a-831e-5b8b0a340b26).
DEBUG 2023-01-12 18:45:03,348 [TCP Receiver Thread on Notes (f8bb636c-4414-4a9a-831e-5b8b0a340b26) < pool-38-thread-1] com.mirth.connect.connectors.tcp.TcpReceiver: Reading from socket input stream (TCP Listener "Source" on channel f8bb636c-4414-4a9a-831e-5b8b0a340b26)...
DEBUG 2023-01-12 18:45:03,548 [TCP Receiver Thread on Notes (f8bb636c-4414-4a9a-831e-5b8b0a340b26) < pool-38-thread-1] com.mirth.connect.connectors.tcp.TcpReceiver: Timeout reading from socket input stream (TCP Listener "Source" on channel f8bb636c-4414-4a9a-831e-5b8b0a340b26).
DEBUG 2023-01-12 18:45:03,549 [TCP Receiver Thread on Notes (f8bb636c-4414-4a9a-831e-5b8b0a340b26) < pool-38-thread-1] com.mirth.connect.connectors.tcp.TcpReceiver: Reading from socket input stream (TCP Listener "Source" on channel f8bb636c-4414-4a9a-831e-5b8b0a340b26)...
DEBUG 2023-01-12 18:45:03,628 [TCP Receiver Thread on Notes (f8bb636c-4414-4a9a-831e-5b8b0a340b26) < pool-38-thread-1] com.mirth.connect.connectors.tcp.TcpReceiver: Bytes returned from socket, length: 53206 (TCP Listener "Source" on channel f8bb636c-4414-4a9a-831e-5b8b0a340b26)
DEBUG 2023-01-12 18:45:03,633 [TCP Receiver Thread on Notes (f8bb636c-4414-4a9a-831e-5b8b0a340b26) < pool-38-thread-1] com.mirth.connect.connectors.tcp.TcpReceiver: Done with socket input stream (TCP Listener "Source" on channel f8bb636c-4414-4a9a-831e-5b8b0a340b26).
DEBUG 2023-01-12 18:45:03,705 [TCP Receiver Thread on Notes (f8bb636c-4414-4a9a-831e-5b8b0a340b26) < pool-38-thread-1] com.mirth.connect.connectors.tcp.TcpReceiver: Reading from socket input stream (TCP Listener "Source" on channel f8bb636c-4414-4a9a-831e-5b8b0a340b26)...
DEBUG 2023-01-12 18:45:03,705 [TCP Receiver Thread on Notes (f8bb636c-4414-4a9a-831e-5b8b0a340b26) < pool-38-thread-1] com.mirth.connect.connectors.tcp.TcpReceiver: Bytes returned from socket, length: 8372 (TCP Listener "Source" on channel f8bb636c-4414-4a9a-831e-5b8b0a340b26)
DEBUG 2023-01-12 18:45:03,708 [TCP Receiver Thread on Notes (f8bb636c-4414-4a9a-831e-5b8b0a340b26) < pool-38-thread-1] com.mirth.connect.connectors.tcp.TcpReceiver: Done with socket input stream (TCP Listener "Source" on channel f8bb636c-4414-4a9a-831e-5b8b0a340b26).
DEBUG 2023-01-12 18:45:03,810 [TCP Receiver Thread on Notes (f8bb636c-4414-4a9a-831e-5b8b0a340b26) < pool-38-thread-1] com.mirth.connect.connectors.tcp.TcpReceiver: Reading from socket input stream (TCP Listener "Source" on channel f8bb636c-4414-4a9a-831e-5b8b0a340b26)...
DEBUG 2023-01-12 18:45:04,010 [TCP Receiver Thread on Notes (f8bb636c-4414-4a9a-831e-5b8b0a340b26) < pool-38-thread-1] com.mirth.connect.connectors.tcp.TcpReceiver: Timeout reading from socket input stream (TCP Listener "Source" on channel f8bb636c-4414-4a9a-831e-5b8b0a340b26).
```

There does not seem to be a pattern, the stream is reset at various times throughout transmission not after a specific length. Medium size messages seem to succeed after a couple retries it's only really large message that have issues. We've increased buffer size and memory, but neither seems to be working. Current configuration is as follows:

<img width="484" alt="mirth-source-settings" src="https://user-images.githubusercontent.com/39662961/212974078-3bcf7766-ce7f-4d2a-a064-35630cc1619e.png">

If anyone knows of more DEBUG flags I can set or somewhere else to dig for TCP related logs would love to get more details. Happy to post additional details/captures/configs. Thanks!

**Comment by jonbartels** [comment, marked as answer]
A 200ms receive timeout seems quite low. Does the problem persist even after increasing this timeout?

**Reply by pacmano1** [reply]
I'd say FAR to low :).   I think that calculates to 15Mbs, 3Mb at 200ms or so.   That isn't fast per se, depends on your topology.

**Reply by jonbartels** [reply]
Reading the PCAP (very helpful BTW! Great job with triage @wcameronbowen !) the FIN is damn close to 200ms. 974.81250 to 975.047171 is 230ms. Close enough to 200ms for our purposes.

**Reply by wcameronbowen** [reply]
Testing higher timeout (1000ms) now, will update on how that goes. Any rule of thumb for setting reasonable timeouts? I did a lot of digging to find recommended settings for connections with large message sizes. Not much luck other than ensuring buffer sizes were larger than normal (though no recommended sizes or formula for calculating based on average message size). Thanks all!

**Reply by pacmano1** [reply]
Most of mine are set to 0, aka no timeout.

**Reply by wcameronbowen** [reply]
I have the timeout set to `0` (since that is the default). Thanks a bunch for all the assistance. I think the root issue was that too many channel settings were being changed at once. I set everything, but the buffer size back to default and so far not a single message has failed today (which is rare).
